### PR TITLE
Data structures

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR adds data structures to work with KSUIDs.

The first one is the `Sequence` type, which can be used to generate an ordered sequence of KSUIDs.

The second is the `CompressedSet` type, which is an ordered set of KSUIDs that uses techniques inspired from bitmap compression to efficiently store KSUIDs, especially when IDs are packed in sequences or _close_ to each other.

Here are the test and benchmarks:
```
$ go test -v -run CompressedSet -bench CompressedSet -benchmem
=== RUN   TestCompressedSet
=== RUN   TestCompressedSet/sparse
=== RUN   TestCompressedSet/packed
=== RUN   TestCompressedSet/mixed
--- PASS: TestCompressedSet (0.00s)
    --- PASS: TestCompressedSet/sparse (0.00s)
    	set_test.go:126: original 20000 B, compressed 17102 B (14.49%)
    --- PASS: TestCompressedSet/packed (0.00s)
    	set_test.go:126: original 20000 B, compressed 2153 B (89.23%)
    --- PASS: TestCompressedSet/mixed (0.00s)
    	set_test.go:126: original 20000 B, compressed 5022 B (74.89%)
goos: darwin
goarch: amd64
pkg: github.com/segmentio/ksuid
BenchmarkCompressedSet/write-4         	    3000	    465194 ns/op	      32 B/op	       1 allocs/op
BenchmarkCompressedSet/read-4          	   20000	     62202 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/segmentio/ksuid	3.380s
```

- **465us to compress 1000 KSUIDs** the profile shows that this is largely dominated by the sorting operation, I'll see what I can do about this
- **62us to decompress 1000 KSUIDs** I think this is pretty good perf, especially considering the kind of compression ratios that this data structure can achieve when storing lots of KSUIDs that were produced around the same time or using sequences.

screen shot 2017-09-27 at 4 28 57 pm

Please take a look and let me know if anything should be changed.

### CODE REVIEW

File: sequence.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// Up to 65536 KSUIDs can be generated by for a single seed.
+//
+// A typical usage of a Sequence looks like this:
+//
+//	seq := ksuid.Sequence{
+//		Seed: ksuid.New(),
+//	}
+//	id, err := seq.Next()
+//
+type Sequence struct {
```
Comment (f2prateek) [inline sequence.go]: Should we document the implementation doesn't guarantee thread safety?

Comment (achille-roussel) [reply sequence.go]: Sure, I'll add that.

File: set.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+		// The first KSUID is always written to the set, this is the starting
+		// point for all deltas.
+		set = append(set, ids[0][:]...)
+
+		timestamp := ids[0].Timestamp()
+		lastKSUID := ids[0]
+		lastValue := makeUint128FromPayload(ids[0][timestampLengthInBytes:])
+
+		for i := 1; i != len(ids); i++ {
```
Comment (f2prateek) [inline set.go]: Curios - is it intentional this isn't the more traditional `i < len(ids)` or just a style thing?

File: set.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
+package ksuid
+
+import "encoding/binary"
+
+// CompressedSet is an immutable data type which stores a set of KSUIDs.
+type CompressedSet []byte
+
+// Iter returns an iterator that produces all KSUIDs in the set.
```
Comment (f2prateek) [inline set.go]: What's the common use case of using the iterator? Wondering if most of the time folks would just want to get all the []KSUID in the list.

File: set.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+// AppendCompressed uses the given byte slice as pre-allocated storage space to
+// build a KSUID set.
+//
+// Note that the set uses a compression technique to store the KSUIDs, so the
+// resuling length is not 20 x len(ids). The rule of thumb here is for the given
+// byte slice to reserve the amount of memory that the application would be OK
+// to waste.
+func AppendCompressed(set []byte, ids ...KSUID) CompressedSet {
```
Comment (f2prateek) [inline set.go]: What if the supplied slice is smaller than the amount of memory required to build the compressed set?

File: set.go (hunk 2 above)
Comment (achille-roussel) [reply set.go]: That's just a style thing.

File: set.go (hunk 3 above)
Comment (achille-roussel) [reply set.go]: If you generate a slice to hold the entire set, you kind of defeat the purpose of using a compressed data structure don't you?

File: set.go (hunk 4 above)
Comment (achille-roussel) [reply set.go]: It's and `append`, it'll get extended as needed.

File: set.go (hunk 3 above)
Comment (f2prateek) [reply set.go]: Ah I see, yep that makes sense

**Comment by achille-roussel** [comment]
Thanks for the review! I'm going to vendor this branch for a little while to experiment with this on production load before I merge to master.

File: sequence.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+//		Seed: ksuid.New(),
+//	}
+//	id, err := seq.Next()
+//
+// Sequence values are not safe to use concurrently from multiple goroutines.
+type Sequence struct {
+	// The seed is used as base for the KSUID generator, all generated KSUIDs
+	// share the same leading 18 bytes of the seed.
+	Seed  KSUID
+	count uint32 // uint32 for overlow, only 2 bytes are used
```
Comment (f2prateek) [inline sequence.go]: typo: overflow

**Comment by achille-roussel** [comment]
I spent a bit of time working on reimplementing a quicksort for KSUIDs so we don't pay the cost of the `sort.Interface` abstraction, seems like it doubles the throughput on building the compressed set of KSUIDs.
```
$ go test -v -run CompressedSet -bench CompressedSet
=== RUN   TestCompressedSet
=== RUN   TestCompressedSet/sparse
=== RUN   TestCompressedSet/packed
=== RUN   TestCompressedSet/mixed
--- PASS: TestCompressedSet (0.00s)
    --- PASS: TestCompressedSet/sparse (0.00s)
    	set_test.go:126: original 20000 B, compressed 17102 B (14.49%)
    --- PASS: TestCompressedSet/packed (0.00s)
    	set_test.go:126: original 20000 B, compressed 2153 B (89.23%)
    --- PASS: TestCompressedSet/mixed (0.00s)
    	set_test.go:126: original 20000 B, compressed 5022 B (74.89%)
goos: darwin
goarch: amd64
pkg: github.com/segmentio/ksuid
BenchmarkCompressedSet/write-4         	   10000	    238968 ns/op
BenchmarkCompressedSet/read-4          	   20000	     61904 ns/op
PASS
ok  	github.com/segmentio/ksuid	4.298s
```

**Comment by achille-roussel** [comment]
For completeness I also put together a benchmark for comparing the performance of gzip and snappy => https://gist.github.com/achille-roussel/70d7b76be2e46b12d7acf6b463ce757e

Pretty encouraging to see that the compressed set performs better than gzip while being 5 to 10x faster. Snappy still provides more throughput but the compression ratios are way lower, plus encoding and decoding cannot be done in place with either compression algorithms.

**Comment by achille-roussel** [comment]
I noticed there was an issue, range compression wasn't done properly, it is now... results are really massive (99% compression ratio 😉  for really high density distributions)

**Comment by achille-roussel** [comment]
This has been used for a couple of days in production and is showing great results, I'll merge so we can start pointing the dependency back at master.

**Pull request merged**